From b0e66da012ab53fe2f0fb60018b7a434e26d2e3a Mon Sep 17 00:00:00 2001
From: xtex <xtex@astrafall.org>
Date: Sun, 8 Mar 2026 09:40:23 +0800
Subject: [PATCH 128/133] std.zig.system.loongarch: take HWCAP into
 consideration
X-Developer-Signature: v=1; a=openpgp-sha256; l=2385; i=xtex@astrafall.org;
 h=from:subject; bh=Bjfqv09c4KMNdlDWhYJ3RqEuK8kX2iBauefXqchIBUQ=;
 b=kA0DAAoWuRgIbtgEW5EByyZiAGm+bhDIr2WZ0t1JF692xcedRvk4rlj1YvW8VbBwah9lW1SEp
 YiRBAAWCgA5FiEEcjGASwUsZw8VpncduRgIbtgEW5EFAmm+bhAbFIAAAAAABAAObWFudTIsMi41
 KzEuMTEsMiwyAAoJELkYCG7YBFuREscA/3xCaFxYs7Ur3fuUYktLeciD+DL5xTJzZwtYJKYDVBb
 jAP4rlpslWyhXMGKZw+985eJglfIvNMRTr5g5cqRnVfLQBw==
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

- CPUCFG is not available on LA32R,
  although LA32R is not a real thing yet
- Prefer information from HWCAP than CPUCFG.
  Assume features unavailable if HWCAP is not available.
  This is because many features require kernel support
  (e.g. context switching support).

Change-Id: I0a610b2c4fc751f0f159e79e01bb9f6fb7588ccf
Signed-off-by: xtex <xtex@astrafall.org>
---
 lib/std/zig/system/loongarch.zig | 25 +++++++++++++++++--------
 1 file changed, 17 insertions(+), 8 deletions(-)

diff --git a/lib/std/zig/system/loongarch.zig b/lib/std/zig/system/loongarch.zig
index 2186ce6c76a2..30d48a725138 100644
--- a/lib/std/zig/system/loongarch.zig
+++ b/lib/std/zig/system/loongarch.zig
@@ -1,6 +1,9 @@
 const builtin = @import("builtin");
 const std = @import("std");
 
+const host_has_hwcap = builtin.os.tag == .linux;
+const HWCAP = if (host_has_hwcap) std.os.linux.HWCAP else undefined;
+
 inline fn bit(input: u32, offset: u5) bool {
     return (input >> offset) & 1 != 0;
 }
@@ -27,19 +30,25 @@ pub fn detectNativeCpuAndFeatures(
 
     cpu.features.addFeatureSet(cpu.model.features);
 
-    const cfg1 = cpucfg(1);
-    const cfg2 = cpucfg(2);
-    const cfg3 = cpucfg(3);
+    const hwcap_bits: usize = if (host_has_hwcap) std.os.linux.getauxval(std.elf.AT_HWCAP) else 0;
+    // CPUCFG is not available in LA32R, although LA32R is not a real thing yet.
+    const has_cpucfg = if (host_has_hwcap)
+        (hwcap_bits & HWCAP.CPUCFG) != 0
+    else
+        builtin.cpu.has(.loongarch, .@"64bit");
 
-    setFeature(&cpu, .ual, bit(cfg1, 20));
+    setFeature(&cpu, .ual, (hwcap_bits & HWCAP.UAL) != 0);
 
-    const has_fpu = bit(cfg2, 0);
+    const cfg2 = if (has_cpucfg) cpucfg(2) else 0;
+    const cfg3 = if (has_cpucfg) cpucfg(3) else 0;
+
+    const has_fpu = (hwcap_bits & HWCAP.FPU) != 0;
     setFeature(&cpu, .f, has_fpu and bit(cfg2, 1));
     setFeature(&cpu, .d, has_fpu and bit(cfg2, 2));
 
-    setFeature(&cpu, .lsx, bit(cfg2, 6));
-    setFeature(&cpu, .lasx, bit(cfg2, 7));
-    setFeature(&cpu, .lvz, bit(cfg2, 10));
+    setFeature(&cpu, .lsx, (hwcap_bits & HWCAP.LSX) != 0);
+    setFeature(&cpu, .lasx, (hwcap_bits & HWCAP.LASX) != 0);
+    setFeature(&cpu, .lvz, (hwcap_bits & HWCAP.LVZ) != 0);
 
     setFeature(&cpu, .lbt, bit(cfg2, 18) and bit(cfg2, 19) and bit(cfg2, 20));
 
-- 
2.52.0

